typed-racket-lib/typed-racket/typecheck/tc-lambda-unit.rkt

```diff
@@ -217,11 +236,31 @@
           (get-type rest-id #:default Univ)]
          ;; Lambda with no rest argument
          [else #f]))
-
-     (list
-       (tc-lambda-body arg-list arg-types
-         #:rest (and rest-type (list rest-id rest-type))
-         body))]))
+     (cond 
+      ;; special case for un-annotated eta-expansions
+      [(and eta-expand? (not rest-id) (andmap not arg-types)
+            ;; FIXME: should also handle polymorphic types
+            ;; but we can't return anything but a (listof arr?) here 
+            ;; FIXME: sometimes will typecheck the relevant code twice if it doesn't match
+            ;; FIXME: misses optimization opportunities of this code
+            (match (tc-expr eta-expand?)
+              [(tc-result1: (Function: arrs))
+               (define possibles (for*/list ([arr (in-list arrs)]
+                                             [restricted (in-value (restrict-to-arity arr (length arg-list)))]
+                                             #:when restricted)
+                                   restricted))
+               (if (null? possibles)
+                   #f
+                   possibles)]))
+       =>
+       (lambda (x)
+         (register-ignored! (car (syntax-e body)))
```

Comment:
Why ignore the body? Don't you have all the info required to add types to the type table?